### PR TITLE
Convert docker-based action to typescript-based one

### LINKED ISSUE #11: The container isn't terminated when the action run in self-hosted runner
Hello,

Now I'm trying to use this action in our self-hosted runner (Ubuntu 18.04), and found that this action doesn't terminate the docker instance even after the action run has finished.
I [tried to use `post-entrypoint`](https://github.com/saucelabs/sauce-connect-action/compare/master...KengoTODA:shutdown-sauce-connect) to run `docker stop`, but it failed because the cleanup operation runs in another docker instance then it cannot touch the instance runs `saucelabs/sauce-connect`.

To solve this issue from root, I think we need to migrate this action to [the JavaScript based action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-javascript-action), and call `docker` command from both [`main`](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/metadata-syntax-for-github-actions#runsmain) and [`post`](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/metadata-syntax-for-github-actions#post) then the operation triggered by `post` can stop the instance by `docker stop`.

I'm not sure why we choose docker-based implementation, so could you tell me how we should solve this issue?
If you're welcome to see PR for this conversion, I will handle it. Thanks in advance!

note: Known workaround is terminate them manually at https://app.saucelabs.com/tunnels

### PR DESCRIPTION
**Pull request by KengoTODA**
To close #11, convert the docker-based action to typescript-based one, then we can use [`post`](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/metadata-syntax-for-github-actions#post) to run the cleanup operation.

Note that the configurations for build tools are copied from [the official template project](https://github.com/actions/typescript-action). I hope that it reduces the barrier to contribute for other users.

### CODE REVIEW

File: .prettierrc.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "printWidth": 80,
+    "tabWidth": 4,
```
Comment (KengoTODA) [inline .prettierrc.json:3]: respected the existing config in the `.editorconfig`

File: src/main.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                        }
+                    }
+                }
+            )
+            saveState('containerId', containerId.trim())
+            setTimeout(() => {
+                // 30 seconds is generally enough for Sauce Connect to start
+                info('SC ready')
+                resolve(void 0)
+            }, 30 * 1000)
```
Comment (christian-bromann) [inline src/main.ts:70]: How about we actually check the log file to see if it contains `you may start your tests`. We could either mount the log directory to the GitHub action or just run docker commands to get access to the file.

File: src/main.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import {getInput, saveState, setFailed} from '@actions/core'
+import {exec} from '@actions/exec'
+import {info} from 'console'
+import optionMappingJson from './option-mapping.json'
+
+const CONTAINER_VERSION = '4.6.2'
+const LOG_FILE = '/srv/sauce-connect.log'
```
Comment (christian-bromann) [inline src/main.ts:7]: Is it possible to upload this log file as asset?

File: src/post.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+async function run(): Promise<void> {
+    try {
+        const containerId = getState('containerId')
+        if (!containerId) {
+            warning(
+                'No state found. Assume that no container run in this workflow run.'
+            )
+            return
+        }
```
Comment (christian-bromann) [inline src/post.ts:12]: Let's have this part outside of the `try` block

**Review by christian-bromann — COMMENTED** [review]
This is great! A couple of comments.

File: src/main.ts (hunk 3 above)
Comment (KengoTODA) [reply src/main.ts:7]: possible, refs https://github.com/actions/toolkit/tree/main/packages/artifact

Is it OK to make this change in this PR? I assume that we can handle this new feature in another PR.

Comment (christian-bromann) [reply src/main.ts:7]: > I assume that we can handle this new feature in another PR.

Feel free to add it as part of this PR. If not, let's create an issue for it an tackle it afterwards.

File: src/main.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
+import {getInput, info, saveState, setFailed} from '@actions/core'
+import {exec} from '@actions/exec'
+import {join} from 'path'
+import {tmpdir} from 'os'
+import {promises} from 'fs'
+import {wait} from './wait'
+import optionMappingJson from './option-mapping.json'
+
+const CONTAINER_VERSION = '4.6.2'
```
Comment (christian-bromann) [inline src/main.ts:9]: Making this an GH Action option would be great. Let's do this in a separate PR though.

**Review by christian-bromann — COMMENTED** [review]
Does the way this action is used changed in any way? If so we should update the README accordingly and release with a new major version.

**Comment by KengoTODA** [comment]
> Does the way this action is used changed in any way? If so we should update the README accordingly and release with a new major version.

Input, output and behavior are not changed. Just fix #11. Performance could be improved.

**Review by christian-bromann — APPROVED** [review]
Looks good! Thank you so much 👍

File: package.json [hunk 6]
```diff
@@ -17,18 +17,34 @@
...
   ],
   "scripts": {
-    "docker": "run-s docker:build docker:tag docker:push:*",
-    "docker:build": "docker build -t saucelabs/sauce-connect-action .",
-    "docker:push:latest": "docker push saucelabs/sauce-connect-action:latest",
-    "docker:push:version": "docker push saucelabs/sauce-connect-action:$npm_package_version",
-    "docker:tag": "docker tag saucelabs/sauce-connect-action saucelabs/sauce-connect-action:$npm_package_version",
-    "test": "run-p test:*",
-    "test:setup": "http-server ./test -p 8080 > /dev/null 2>&1 &",
-    "test:run": "node ./test/test.js"
```
Comment (christian-bromann) [inline package.json:27]: Could we get back this test?

**Review by christian-bromann — COMMENTED** [review]
I think I prematurely merged this PR. Can you revert the test command to also execute our e2e test?

File: package.json (hunk 6 above)
Comment (christian-bromann) [reply package.json:27]: https://github.com/saucelabs/sauce-connect-action/issues/15

**Comment by KengoTODA** [comment]
Can you reproduce the issue? 
In my environment, this implementation works without this problem. It'll be a great help to solve this issue for other users. :)
